### PR TITLE
Add smpc argmax

### LINKED ISSUE #133: Add implementation of Max / Argmax
## Feature Description
We used to have them in syft 0.2:
max: https://github.com/OpenMined/PySyft/blob/syft_0.2.x/syft/frameworks/torch/tensors/interpreters/additive_shared.py#L1089
argmax: https://github.com/OpenMined/PySyft/blob/syft_0.2.x/syft/frameworks/torch/tensors/interpreters/additive_shared.py#L1044

The theory behind is briefly sketched in Section 4.2 in https://arxiv.org/pdf/2006.04593.pdf

Having this back in SyMPC would be so much useful for many usecases 😄

### PR DESCRIPTION
**Pull request by NiWaRe**
## Description
This implements the `argmax` operation on a `MPCTensor`. 
Closes #133 .

## How has this been tested?
- Added unit tests

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Comment by NiWaRe** [comment]
Still fixing some errors, especially with `parallel_execution`.

File: src/sympc/api.py [hunk 1]
```diff
@@ -117,4 +118,6 @@
     ),
     ("sympc.tensor.ShareTensor.unsqueeze", "sympc.tensor.ShareTensor"),
     ("sympc.tensor.ShareTensor.view", "sympc.tensor.ShareTensor"),
+    # TODO: correct way to define a static function?
+    ("sympc.tensor.static.helper_argmax_pairwise", "sympc.tensor.ShareTensor"),
```
Comment (gmuraru) [inline src/sympc/api.py:122]: You need to add:
`"sympc.tensor.static", sympc.tensor.static`
in the `allowed_external_modules`  (another list above this one)

Comment (NiWaRe) [reply src/sympc/api.py:122]: Thanks for the answer! Unfortunately, I get the same error when doing this. @gmuraru

**Review by NiWaRe — COMMENTED** [review]
Hey @gmuraru thanks for changing the call to `parallel_execution` I think I understood it better with the wrapper now. However, if I call `x.argmax(dim=0)` or `x.argmax()` in an extra cell in the examples/introduction.ipynb (just as a quick test), I still get the same error. 
Did the parallel_execution work for you? 

`---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-10-19de3b86437f> in <module>
      1 # x.argmax(dim=0) doesn't yet work -- same error as originally sent to George
----> 2 x.argmax()

~/Documents/AI/PP-ML/SyMPC/src/sympc/tensor/static.py in argmax(self, dim, keepdim, one_hot)
     25     #   results in creating two matrices and substracting them
     26     args = [[share_ptr_tensor, dim] for share_ptr_tensor in self.share_ptrs]
---> 27     shares = parallel_execution(helper_argmax_pairwise, self.session.parties)(args)
     28     # then create an MPCTensor tensor based on this results per share
     29     # (we can do that bc substraction can be done in mpc fashion out of the box)

~/Documents/AI/PP-ML/SyMPC/src/sympc/utils/utils.py in wrapper(args, kwargs)
    107             func_name = f"{fn.__module__}.{fn.__qualname__}"
    108             attr_getter = operator.attrgetter(func_name)
--> 109             funcs = [attr_getter(party) for party in parties]
    110         else:
    111             funcs = list(repeat(fn, nr_parties))

~/Documents/AI/PP-ML/SyMPC/src/sympc/utils/utils.py in <listcomp>(.0)
    107             func_name = f"{fn.__module__}.{fn.__qualname__}"
    108             attr_getter = operator.attrgetter(func_name)
--> 109             funcs = [attr_getter(party) for party in parties]
    110         else:
    111             funcs = list(repeat(fn, nr_parties))

~/Documents/AI/PP-ML/PySyft/src/syft/ast/module.py in __getattribute__(self, item)
    212 
    213     def __getattribute__(self, item: str) -> Any:
--> 214         target_object = super().__getattribute__(item)
    215         if isinstance(target_object, ast.static_attr.StaticAttribute):
    216             return target_object.get_remote_value()

AttributeError: 'Module' object has no attribute 'static'`

File: src/sympc/tensor/static.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+# third party
+import torch
+
+from sympc.utils import parallel_execution
+
+from .mpc_tensor import *  # why doesn't from .mpc_tensor import MPCTensor work?
```
Comment (Syzygianinfern0) [inline src/sympc/tensor/static.py:6]: That is due to [circular/cyclic importing](https://stackoverflow.com/a/746067/8878627)

Comment (NiWaRe) [reply src/sympc/tensor/static.py]: Yep, thanks also noticed that still didn't work for me though. (the remote call).

File: src/sympc/tensor/static.py [hunk 3]
```diff
@@ -11,18 +15,18 @@ def argmax(self, dim=None, keepdim=False, one_hot=False) -> "MPCTensor":
...
     x = self.flatten() if dim is None and len(self.shape) > 1 else self
-    # for each share in MPCTensor 
+    # for each share in MPCTensor
     #   do the algorithm protrayed in paper (helper_argmax_pairwise)
-    #   results in creating two matrices and substracting them 
-    args = [dim]
-    shares = parallel_execution(helper_argmax_pairwise, self.session.parties)(args)    
-    # then create an MPCTensor tensor based on this results per share 
+    #   results in creating two matrices and substracting them
+    args = [[share_ptr_tensor, dim] for share_ptr_tensor in self.share_ptrs]
```
Comment (NiWaRe) [inline src/sympc/tensor/static.py]: Did removing the cyclic import and adding this line already solve the issue for you @gmuraru concerning the remote call? (That's also the version I tried it locally and it didn't work -- is it a problem on my end maybe?)

**Comment by Syzygianinfern0** [comment]
View file changes [here](https://github.com/Syzygianinfern0/SyMPC/compare/feature/cheby_sigmoid...NiWaRe:main) till #136 is merged.

File: src/sympc/api.py [hunk 4]
```diff
@@ -60,29 +60,7 @@
...
-        "sympc.tensor.ShareTensor",
-    ),
-    (
-        "sympc.store.CryptoStore.populate_store",
-        "syft.lib.python._SyNone",
-    ),
-    (
-        "sympc.utils.get_new_generator",
-        "torch.Generator",
-    ),
```
Comment (Syzygianinfern0) [inline src/sympc/api.py:85]: Why are these moved?

Comment (gmuraru) [reply src/sympc/api.py:85]: They are added in a separate list.
I split the `allowed_external_attrs` into 3 lists:
- `allowed_external_attrs`
- `share_tensor_attrs`
- `replicated_tensor_attrs`

File: src/sympc/api.py [hunk 5]
```diff
@@ -135,9 +113,22 @@
...
+    ("sympc.tensor.static.helper_argmax_pairwise", "sympc.tensor.ShareTensor"),
     ("sympc.tensor.ShareTensor.reshape", "sympc.tensor.ShareTensor"),
+    ("sympc.tensor.ShareTensor.repeat", "sympc.tensor.ShareTensor"),
+    ("sympc.tensor.ShareTensor.narrow", "sympc.tensor.ShareTensor"),
+    ("sympc.tensor.ShareTensor.dim", "syft.lib.python.Int"),
+    ("sympc.tensor.ShareTensor.transpose", "sympc.tensor.ShareTensor"),
+    ("sympc.tensor.ShareTensor.flatten", "sympc.tensor.ShareTensor"),
+    ("sympc.tensor.ShareTensor.expand", "sympc.tensor.ShareTensor"),
+    ("sympc.tensor.ShareTensor.roll", "sympc.tensor.ShareTensor"),
+    ("sympc.tensor.ShareTensor.shape", "syft.lib.python.Tuple"),
```
Comment (Syzygianinfern0) [inline src/sympc/api.py:128]: ### Part 1
Bug: `sympc.tensor.ShareTensor.expand` is added twice

### Part 2.1
Question: The other ones (`narrow`, `dim`, ... , `shape`) are to be added ... (see Part 2.2)

File: src/sympc/tensor/mpc_tensor.py [hunk 6]
```diff
@@ -33,6 +33,7 @@
     "squeeze",
     "unsqueeze",
     "view",
+    "expand",
```
Comment (Syzygianinfern0) [inline src/sympc/tensor/mpc_tensor.py:36]: ### Part 2.2
... to be added here as well right?

File: src/sympc/tensor/mpc_tensor.py [hunk 7]
```diff
@@ -118,6 +119,7 @@ class MPCTensor(metaclass=SyMPCTensor):
         "squeeze",
         "unsqueeze",
         "view",
+        "expand",
```
Comment (Syzygianinfern0) [inline src/sympc/tensor/mpc_tensor.py:122]: ### Part 2.3
... and here

File: src/sympc/tensor/mpc_tensor.py [hunk 8]
```diff
@@ -827,6 +829,9 @@ def __getattribute__(self, attr_name: str) -> Any:
         from sympc.tensor.grads import GRAD_FUNCS
         from sympc.tensor.static import STATIC_FUNCS
 
+        if attr_name in STATIC_FUNCS.keys():
+            return functools.partial(STATIC_FUNCS[attr_name], self)
+
         if attr_name in STATIC_FUNCS.keys():
             return functools.partial(STATIC_FUNCS[attr_name], self)
```
Comment (Syzygianinfern0) [inline src/sympc/tensor/mpc_tensor.py:836]: ### Part 3
Bug: These lines are repeated

File: src/sympc/tensor/share_tensor.py [hunk 9]
```diff
@@ -27,6 +27,7 @@
     "squeeze",
     "unsqueeze",
     "view",
+    "expand",
```
Comment (Syzygianinfern0) [inline src/sympc/tensor/share_tensor.py:30]: ### Part 2.4
... and here

File: src/sympc/tensor/share_tensor.py [hunk 10]
```diff
@@ -36,6 +37,7 @@
     "narrow",
     "dim",
     "transpose",
+    "roll",
```
Comment (Syzygianinfern0) [inline src/sympc/tensor/share_tensor.py:40]: ### Part 2.5
... and here

File: src/sympc/tensor/share_tensor.py [hunk 11]
```diff
@@ -81,6 +83,7 @@ class ShareTensor(metaclass=SyMPCTensor):
         "unsqueeze",
         "t",
         "view",
+        "expand",
```
Comment (Syzygianinfern0) [inline src/sympc/tensor/share_tensor.py:86]: ### Part 2.6
... and here

File: src/sympc/tensor/share_tensor.py [hunk 12]
```diff
@@ -89,6 +92,7 @@ class ShareTensor(metaclass=SyMPCTensor):
         "narrow",
         "dim",
         "transpose",
+        "roll",
```
Comment (Syzygianinfern0) [inline src/sympc/tensor/share_tensor.py:95]: ### Part 2.7
.... and here

File: src/sympc/tensor/static.py [hunk 13]
```diff
@@ -119,4 +121,170 @@ def cat_share_tensor(session_uuid_str: str, *shares: Tuple[ShareTensor]) -> Shar
...
+        argmax_mpc = argmax_mpc * torch.Tensor([i for i in range(shape[dim])]).view(
+            size
+        )
+
+        max_mpc = max_mpc.sum(dim=dim, keepdim=keepdim)
+        res = max_mpc, argmax_mpc.sum(dim=dim, keepdim=keepdim)
+    return res
+
+
+# from syft < 0.3.0
```
Comment (Syzygianinfern0) [inline src/sympc/tensor/static.py:253]: ### Part 4
Maybe: This can be a part of the docstring like how CrypTen is acknowledged in the `def argmax`

**Review by Syzygianinfern0 — CHANGES_REQUESTED** [review]
I have no clue if the comments appear in the order I want them to. So, I have labelled them as parts 😬

File: src/sympc/api.py (hunk 5 above)
Comment (gmuraru) [reply src/sympc/api.py:128]: I think we can remove those. Since we are not using them remotely, only locally.

File: src/sympc/tensor/mpc_tensor.py (hunk 6 above)
Comment (gmuraru) [reply src/sympc/tensor/mpc_tensor.py:36]: Hmm, they were already added - a few lines below, but we don't need them in the api.py, because those are used only if you want to call something remotely, but we call them locally

**Review by Syzygianinfern0 — APPROVED** [review]
🚀
